File: Data Structures & Algorithms/three-integer-sum/submission-1.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end()) ;
        vector<vector<int>> ans ;
        set<vector<int>> s ;
        for(int i = 0;i<nums.size();i++){
            int j=i+1,k=nums.size()-1 ;
            while(j < k){
                int x = nums[i] + nums[j] + nums[k] ;
                vector<int> p = {nums[i],nums[j],nums[k]} ;
                if(x == 0 && !s.count(p)){
                    s.insert(p) ;
                    ans.push_back(p) ;
                    j++ ;
                    k-- ;
                }
                else if(x < 0){
                    j++ ;
                }
                else {
                    k-- ;
                }
            }    
        }

        return ans ;
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-1.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        
        self.st = set()
        n = len(nums)
        nums.sort()
        def dp(i, curr):
            if i == n:
                self.st.add(tuple(curr))
                return
            curr.append(nums[i])
            dp(i+1, curr)
            curr.pop()
            while i+1 < n and nums[i] == nums[i+1]:
                i += 1
            dp(i+1, curr)
            return

        dp(0,[])

        return list(self.st)
